scripts: tests: sanitycheck: Add more testcases for Class TestSuite

aasthagr opened the pull request:
test_testsuite_class.py: Added testcases for add_configurations(),
get_all_testcases(), get_toolchain() and get_platforms() functions
test_data/board_config: board root directory for all board
configurations. This is also to add the platform configurations
in a list which is used for further testing.
conftest.py: Added fixtures for all_testcases_dict (returns a
dictionary of testcases) and platforms_list (returns a list of
all platforms using add_configurations function).

Signed-off-by: Aastha Grover <aastha.grover@intel.com>

aasthagr committed:
scripts: tests: sanitycheck: Add more testcases for Class TestSuite

test_testsuite_class.py: Added testcases for add_configurations(),
get_all_testcases(), get_toolchain() and get_platforms() functions
test_data/board_config: board root directory for all board
configurations. This is also to add the platform configurations
in a list which is used for further testing.
conftest.py: Added fixtures for all_testcases_dict (returns a
dictionary of testcases) and platforms_list (returns a list of
all platforms using add_configurations function).

Signed-off-by: Aastha Grover <aastha.grover@intel.com>

nashif commented on scripts/tests/sanitycheck/test_testsuite_class.py:
```diff
@@ -38,3 +39,44 @@ def test_testsuite_add_testcases(class_testsuite):
+    Test : Monkeypatch to delete the  ZEPHYR_TOOLCHAIN_VARIANT env var
+    and check if appropriate error is raised"""
+    toolchain = class_testsuite.get_toolchain()
+    assert toolchain in ["zephyr", "gnuarmemb"]
```
here you are assuming that ZEPHYR_TOOLCHAIN_VARIANT is already set in the environment and is zephyr or gnuarmemb, this might fail if someone  has something else set or if the variable is not set.

aasthagr replied:
Removed the assumption that ZEPHYR_TOOLCHAIN_VARIANT can only be zephyr or gnuarmemb. This is now only testing for the error message if ZEPHYR_TOOLCHAIN_VARIANT is not in the environment. But testcase wont fail if it is not in env while running the testcases.

nashif replied:
the solution is not to delete assert completely, you could still set the environment there and keep the test, just do not rely on user environment.

aasthagr replied:
So get_toolchain() function reads the ZEPHYR_TOOLCHAIN_VARIANT from user environment and returns that toolchain. If I keep that test scenario then it will have to rely on user environment. Am i missing any info here ?

nashif replied:
you can mock the env. variable

aasthagr replied:
Fixed this and added the test scenario back by mocking the env. Thanks.

aasthagr committed:
tests: sanitycheck: Testcases for load_from_file and apply_filters

test_testsuite_class.py: Added testcases for load_From_file and
apply_filters functions of Class Testsuite.
conftest.py: Modified class_testsuite fixture to create the
outdir as temporary directory which gets deleted after execution
of testcases.
test_data/sanitycheck.csv: load_from_file function uses this
customized file to load the failed or last run testcases.
test_data/sanitycheck_keyerror.csv: file used by
test_load_from_file function to raise the appropriate error
if there is a keyerror.
scripts/requirements-build-test.txt: added mock & csv python libraries

Signed-off-by: Aastha Grover <aastha.grover@intel.com>

nashif commented:
this is currently failing for me:
```
E               FileNotFoundError: [Errno 2] No such file or directory: '/home/nashif/Work/zephyrproject/zephyr/scripts/tests/sanitycheck/test_data/sanity_out_demo'

/usr/lib64/python3.8/shutil.py:704: FileNotFoundError
------------------------------------------------------------------- Captured log setup -------------------------------------------------------------------
DEBUG    sanitycheck:sanitylib.py:2500 Reading test case configuration files under /home/nashif/Work/zephyrproject/zephyr/scripts/tests/sanitycheck/test_data/testcases/tests...
DEBUG    sanitycheck:sanitylib.py:2503 scanning /home/nashif/Work/zephyrproject/zephyr/scripts/tests/sanitycheck/test_data/testcases/tests
DEBUG    sanitycheck:sanitylib.py:2503 scanning /home/nashif/Work/zephyrproject/zephyr/scripts/tests/sanitycheck/test_data/testcases/tests/test_a
DEBUG    sanitycheck:sanitylib.py:2511 Found possible test case in /home/nashif/Work/zephyrproject/zephyr/scripts/tests/sanitycheck/test_data/testcases/tests/test_a
DEBUG    sanitycheck:sanitylib.py:2503 scanning /home/nashif/Work/zephyrproject/zephyr/scripts/tests/sanitycheck/test_data/testcases/tests/test_c
DEBUG    sanitycheck:sanitylib.py:2511 Found possible test case in /home/nashif/Work/zephyrproject/zephyr/scripts/tests/sanitycheck/test_data/testcases/tests/test_c
DEBUG    sanitycheck:sanitylib.py:2503 scanning /home/nashif/Work/zephyrproject/zephyr/scripts/tests/sanitycheck/test_data/testcases/tests/test_b
DEBUG    sanitycheck:sanitylib.py:2511 Found possible test case in /home/nashif/Work/zephyrproject/zephyr/scripts/tests/sanitycheck/test_data/testcases/tests/test_b
DEBUG    sanitycheck:sanitylib.py:2500 Reading test case configuration files under /home/nashif/Work/zephyrproject/zephyr/scripts/tests/sanitycheck/test_data/testcases/samples...
DEBUG    sanitycheck:sanitylib.py:2503 scanning /home/nashif/Work/zephyrproject/zephyr/scripts/tests/sanitycheck/test_data/testcases/samples
DEBUG    sanitycheck:sanitylib.py:2503 scanning /home/nashif/Work/zephyrproject/zephyr/scripts/tests/sanitycheck/test_data/testcases/samples/test_app
DEBUG    sanitycheck:sanitylib.py:2511 Found possible test case in /home/nashif/Work/zephyrproject/zephyr/scripts/tests/sanitycheck/test_data/testcases/samples/test_app
DEBUG    sanitycheck:sanitylib.py:2453 Reading platform configuration files under /home/nashif/Work/zephyrproject/zephyr/scripts/tests/sanitycheck/test_data/board_config...
DEBUG    sanitycheck:sanitylib.py:2457 Found platform configuration /home/nashif/Work/zephyrproject/zephyr/scripts/tests/sanitycheck/test_data/board_config/1_level/2_level/board_config_2.yaml
DEBUG    sanitycheck:sanitylib.py:2457 Found platform configuration /home/nashif/Work/zephyrproject/zephyr/scripts/tests/sanitycheck/test_data/board_config/1_level/2_level/board_config_1.yaml
DEBUG    sanitycheck:sanitylib.py:2457 Found platform configuration /home/nashif/Work/zephyrproject/zephyr/scripts/tests/sanitycheck/test_data/board_config/1_level/2_level/board_config_3.yaml
================================================================ short test summary info =================================================================
FAILED scripts/tests/sanitycheck/test_testinstance.py::test_create_overlay[True-True-coverage_platform0-native-\nCONFIG_COVERAGE=y\nCONFIG_COVERAGE_DUMP=y\nCONFIG_ASAN=y]
FAILED scripts/tests/sanitycheck/test_testinstance.py::test_create_overlay[False-True-coverage_platform1-native-\nCONFIG_COVERAGE=y\nCONFIG_COVERAGE_DUMP=y]
FAILED scripts/tests/sanitycheck/test_testinstance.py::test_create_overlay[True-True-coverage_platform2-mcu-\nCONFIG_COVERAGE=y\nCONFIG_COVERAGE_DUMP=y]
FAILED scripts/tests/sanitycheck/test_testinstance.py::test_create_overlay[False-False-coverage_platform3-native-] - FileNotFoundError: [Errno 2] No su...
FAILED scripts/tests/sanitycheck/test_testinstance.py::test_create_overlay[False-True-coverage_platform4-native-] - FileNotFoundError: [Errno 2] No suc...
FAILED scripts/tests/sanitycheck/test_testinstance.py::test_create_overlay[True-False-coverage_platform5-native-\nCONFIG_ASAN=y] - FileNotFoundError: [...
============================================================== 6 failed, 38 passed in 0.64s ==============================================================
```

nashif commented on scripts/tests/sanitycheck/test_testsuite_class.py:
```diff
@@ -3,17 +3,20 @@
 import os
+import csv
+import pytest
+from mock import call, patch, MagicMock
```
This needs to be added to requirements*.txt somewhere, I had to install it after import failed.

aasthagr replied:
Addressed this in requirements-build-test.txt.

nashif commented on scripts/tests/sanitycheck/test_testsuite_class.py:
```diff
@@ -38,3 +41,170 @@ def test_testsuite_add_testcases(class_testsuite):
+    """
+    class_testsuite.platforms = platforms_list
+    class_testsuite.testcases = all_testcases_dict
+    for plat in class_testsuite.platforms:
```
does not look great and does not have as much coverage as we would want to have, for example for ram restrictions, but that is fine for now...

aasthagr replied:
Does the test data line #147 in TESTDATA_PART1 not cover the RAM restriction? Am i missing something here? 

'''
("min_ram", "500", "ram", "256", "Not enough RAM"),
'''

nashif replied:
look at https://github.com/zephyrproject-rtos/zephyr/pull/24518/commits/f88678c13db337fe5c44407af7867212eaa02d39#diff-6fbb35ecd03a71dc2d629502d7d18607R33, the last test in the file as an example...

aasthagr replied:
Added another testcase "test_apply_filters_part3: to test the edge cases for ram and flash values.  Let me know if it looks fine. Thanks

aasthagr commented:
> this is currently failing for me:

This was failing because a testcase teat_create_overlay from Spoorthy's PR had a line "shutil.rmtree(test_data + "sanity_out_demo")". Corrected this already by having the sanity_out_demo directory as the temp directory fixture(gets removed after execution of testcases) in class_testsuite fixture in conftest.py.. 
You can review the PR now. Thanks

nashif commented on scripts/tests/sanitycheck/test_testsuite_class.py:
```diff
@@ -3,17 +3,20 @@
 # SPDX-License-Identifier: Apache-2.0
 # pylint: disable=line-too-long
-# pylint: disable=C0321
+# pylint: disable=multiple-statements
```
the whole idea behind linting is that you follow them, if for some reason you have to disable something, you should add a comment about why this rule does not apply to your change...

nashif reviewed (changes requested):
fix remaining issues please.

aasthagr commented on scripts/requirements-build-test.txt:
```diff
@@ -18,3 +18,9 @@ coverage
+mock
+
+# use for test_data of sanitycheck testsuite in scripts/tests/sanitycheck
+csv
```
Removed csv module from here.

aasthagr commented:
@nashif There is a Kernel testcase which is failing in Shippable and another one is in bluetooth which is failing in buildkite/zephyr. I am assuming this cannot be related to this PR. To confirm, can you please look into it and let me know?

nashif commented:
> @nashif There is a Kernel testcase which is failing in

just ignore it

aasthagr commented:
> > @nashif There is a Kernel testcase which is failing in
> 
> just ignore it

Okay, Thanks. I hope the failed check wont block the merge of this PR.
